lib/cargo.js

```diff
@@ -82,65 +82,131 @@ export function provideBuilder() {
             });
           });
         });
-        return array;
-      };
-      const matchThreadPanic = 'thread \'[^\\\']+\' panicked at \'[^\\\']+\', (?<file>[^\\/][^\\:]+):(?<line>\\d+)';
-      const matchBacktrace = 'at (?<file>[^.\/][^\\/][^\\:]+):(?<line>\\d+)';
+        if (json.code) {
+          trace.push({
+            message: json.code.explanation,
+            type: 'Explanation',
+            severity: 'info'
+          });
+        }
+        json.spans.forEach(span => {
+          messages.push({
+            message: json.message,
+            file: span.file_name,
+            line: span.line_start,
+            line_end: span.line_end,
+            col: span.column_start,
+            col_end: span.column_end,
+            type: json.level, // FIXME: change to `json.code ? json.code : json.level` after https://github.com/steelbrain/linter/issues/1149 is fixed
+            severity: level2severity(json.level),
+            trace: trace
+          });
+        });
+      }
+
+      function matchFunction(output) {
+        const useJson = atom.config.get('build-cargo.jsonErrors');
+        const messages = [];  // resulting collection of high-level messages
+        let msg = null;       // current high-level message (error, warning or panic)
+        let sub = null;       // current submessage (note or help)
+        let match = null;
+
+        output.split(/\n/).forEach(line => {
+          // Cargo final error messages start with 'error:', skip them
+          if (line === null || line === '' || line.substring(0, 6) === 'error:') {
+            msg = null;
+            sub = null;
+          } else if (useJson && line[0] === '{') {
+            // Parse a JSON block
+            parseJsonOutput(line, messages);
+          } else if (line.substring(0, 16) === 'stack backtrace:') {
+            // Start a stacktrace submessage (it's always under a panic message)
+            if (msg !== null) {
+              sub = {
+                message: 'Stack backtrace',
+                type: 'Stack'
+              };
+              msg.trace.push(sub);
+            }
+          } else {
+            // Check for compilation messages
+            match = /^(.+.rs):(\d+):(\d+):(?: (\d+):(\d+))? (error|warning|help|note): (.*)/g.exec(line);
+            if (match) {
+              const type = match[6];
+              const message = match[7];
+              if (type === 'error' || type === 'warning' || msg === null) {
+                msg = {
+                  message: message,
+                  file: match[1],
+                  line: match[2],
+                  line_end: match[4],
+                  col: match[3],
+                  col_end: match[5],
+                  type: type,
+                  trace: []
+                };
+                messages.push(msg);
+                sub = null;
+              } else {
+                sub = {
+                  message: message,
+                  type: 'info'
```

Comment:
if you attach file/line info to notes, you get a nice clickable link inside the message, please add that. Some errors are hard to diagnose from the popup alone without this info